File: app/backend/routers/projects/container.py
import docker
from database import database
from .helpers.containerhelper import create_and_start_container, scaffold_template
from helpers.dockerclient import docker_client
from helpers.structlogger import logger
from .containers.config import project_services_config
from .containers.scaffold import scaffold_import, scaffold_fresh, build_project_groups
from fastapi import HTTPException

async def create_project_container(
    project_id: str,
    project_name: str,
    backend_services=None,
    frontend_services=None,
    db=None,
    import_url=None,
):
    """
    Orchestrates container creation and project scaffolding.
    """
    backend_services = backend_services or []
    frontend_services = frontend_services or []
    db = db or []

    # Fetch service configs
    config = await project_services_config(project_name, backend_services, frontend_services, db)
    configs_map = config["configs_map"]

    # Create and start container
    result = create_and_start_container(
        project_id=project_id,
        project_name=project_name,
        base_tag=config["base_tag"],
        base_type=config["base_type"],
        clean_name=config["clean_name"],
        frontend_port=config["frontend_port"],
        backend_services=backend_services,
        frontend_services=frontend_services,
        db=db,
    )

    container = result["container"]
    scan_result = None

    if import_url:
        scan_result = await scaffold_import(container, import_url)
    else:
        fresh_result= await scaffold_fresh(container, project_name, frontend_services, backend_services, db, configs_map)

    scaffold_template(container, "LoggingService", "/")
    container.exec_run("sh -c 'logd &'", tty=True, detach=True)

    if import_url:
        repo_name = import_url.rstrip("/").split("/")[-1].removesuffix(".git")
        frontend_root = scan_result.frontend_root if scan_result and scan_result.frontend_root else f"/app/workspace/{repo_name}"
        backend_root = scan_result.backend_root if scan_result else None
        db_root = None
        pages = scan_result.pages if scan_result else []
        endpoints = scan_result.endpoints if scan_result else []
    else:
        frontend_root = f"/app/workspace/frontend/{project_name}"
        backend_root = "/app/workspace/backend"
        db_root = "/app/workspace/database"
        pages = fresh_result["pages"]
        endpoints = fresh_result["endpoints"]

    detected_frameworks = [f for f in [
        scan_result.frontend_framework,
        scan_result.backend_framework,
        scan_result.db_framework,
    ] if f] if scan_result else []

    groups = build_project_groups(container, project_name, frontend_services, backend_services, scan_result)

    container.stop()

    await database.execute(
        """
        UPDATE projects
        SET container_id = :container_id
        WHERE project_id = :project_id
        """,
        {"container_id": container.id, "project_id": project_id},
    )

    return {
        **result["metadata"],
        "configs_map": configs_map,
        "scan": scan_result,
        "groups": groups,
        "frontend_root": frontend_root,
        "backend_root": backend_root,
        "db_root": db_root,
        "pages": pages,
        "endpoints": endpoints,
        "detected_frameworks": detected_frameworks,
    }


def get_container(project_id: str) -> docker.models.containers.Container:
    container_name = f"devolib_project_{project_id}"
    return docker_client.containers.get(container_name)



async def delete_project_container(project_id: str):
    volume_name = f"devolib_project_{project_id}"

    try:
        container = get_container(project_id)
        logger.info("Stopping container", project_id=project_id)
        container.stop(timeout=2)
        container.remove()
        logger.info("Container removed", project_id=project_id)
    except docker.errors.NotFound:
        logger.warning("Container not found", project_id=project_id)
    except Exception as e:
        logger.error("Error removing container", project_id=project_id, error=str(e))
        raise

    try:
        volume = docker_client.volumes.get(volume_name)
        volume.remove()
        logger.info("Volume removed", volume_name=volume_name)
    except docker.errors.NotFound:
        logger.warning("Volume not found", volume_name=volume_name)
    except Exception as e:
        logger.warning("Error removing volume", project_id=project_id, error=str(e))




async def start_container(project_id: str) -> docker.models.containers.Container:
    try:
        container = get_container(project_id)
        if container.status != "running":
            container.start()
    except docker.errors.NotFound:
        try:
            container = docker_client.containers.run(
                f"devolib_project_{project_id}",
                name=f"devolib_project_{project_id}",
                network="web",
                detach=True,
                tty=True,
                stdin_open=True,
                command="sh -c 'echo Container started!; tail -f /dev/null'",
            )
        except docker.errors.ImageNotFound:
            raise HTTPException(status_code=404, detail="Docker image not found")

    check = container.exec_run("pgrep logd", tty=False, detach=False)
    if check.exit_code != 0:
        container.exec_run("sh -c 'logd > /var/log/logd.log 2>&1 &'", tty=False, detach=True)
        logger.info("Started logd", project_id=project_id)
    else:
        logger.info("logd already running", project_id=project_id)

    await database.execute(
        "UPDATE projects SET last_online = NOW(), status = 'running' WHERE project_id = :project_id",
        {"project_id": project_id}
    )

    return container




async def stop_running_container(project_id: str) -> docker.models.containers.Container:
    container = get_container(project_id)
    container.stop(timeout=2)
    await database.execute(
        "UPDATE projects SET status = 'stopped' WHERE project_id = :project_id",
        {"project_id": project_id}
    )
    return container

File: backend/handlers/projects/container.go
package handlers

import (
	"context"
	"database/sql"
	"devolib/structs"
	"encoding/json"
	"fmt"
	"path"
	"strings"
)

type CreateProjectResult struct {
	ProjectID          string
	ContainerID        string
	Port               int
	URL                string
	BaseType           string
	CleanName          string
	FrontendRoot       string
	BackendRoot        string
	DBRoot             string
	Pages              []structs.Page
	Endpoints          []structs.Endpoint
	Groups             []structs.ProjectGroup
	DetectedFrameworks []string
}

func compactStrings(values ...string) []string {
	result := make([]string, 0, len(values))
	for _, value := range values {
		if value != "" {
			result = append(result, value)
		}
	}
	return result
}

func scaffoldImport(ctx context.Context, helper *ContainerHelper, containerID, importURL string) (*structs.ScanResult, error) {
	repoName := strings.TrimSuffix(path.Base(strings.TrimSuffix(importURL, "/")), ".git")
	repoPath := "/app/workspace/" + repoName

	helper.Log.Info("cloning repo", "url", importURL)

	if _, err := helper.ExecInContainer(ctx, containerID,
		fmt.Sprintf("cd /app/workspace && git clone --depth=1 %s", shellQuote(importURL))); err != nil {
		return nil, fmt.Errorf("cloning repository: %w", err)
	}

	helper.Log.Info("scanning project structure", "url", importURL)
	scan := ScanProject(ctx, helper, containerID, repoPath)

	if FileExists(ctx, helper, containerID, repoPath+"/package.json") {
		helper.Log.Info("installing frontend dependencies", "url", importURL)
		if _, err := helper.ExecInContainer(ctx, containerID,
			fmt.Sprintf("cd %s && npm install", shellQuote(repoPath))); err != nil {
			helper.Log.Warn("frontend dependency installation failed", "error", err)
		}
	}

	if FileExists(ctx, helper, containerID, repoPath+"/requirements.txt") {
		helper.Log.Info("installing backend dependencies", "url", importURL)
		if _, err := helper.ExecInContainer(ctx, containerID,
			"pip install -r "+shellQuote(repoPath+"/requirements.txt")); err != nil {
			helper.Log.Warn("backend dependency installation failed", "error", err)
		}
	}

	return scan, nil
}

func scaffoldFresh(
	ctx context.Context,
	helper *ContainerHelper,
	containerID, projectName string,
	frontendServices, backendServices, dbServices []string,
	configsMap map[string]ServiceConfig,
) (*structs.ScaffoldResult, error) {
	result := &structs.ScaffoldResult{
		Pages:     []structs.Page{},
		Endpoints: []structs.Endpoint{},
	}

	for _, framework := range frontendServices {
		cfg, ok := configsMap[framework]
		if !ok {
			continue
		}

		if cfg.ScaffoldCommand != nil {
			cmd := strings.ReplaceAll(*cfg.ScaffoldCommand, "{name}", projectName)
			helper.Log.Info("scaffolding frontend", "framework", framework, "cmd", cmd)

			if _, err := helper.ExecInContainer(ctx, containerID,
				fmt.Sprintf("cd /app/workspace/frontend && %s", cmd)); err != nil {
				helper.Log.Warn("frontend scaffold failed", "framework", framework, "error", err)
			}
		}

		if cfg.DefaultPackages != "" && cfg.DefaultPackages != "[]" {
			var packages []string
			if err := json.Unmarshal([]byte(cfg.DefaultPackages), &packages); err != nil {
				helper.Log.Warn("invalid default packages", "framework", framework, "error", err)
			} else if len(packages) > 0 {
				if _, err := helper.ExecInContainer(ctx, containerID,
					fmt.Sprintf("cd %s && npm install %s",
						shellQuote("/app/workspace/frontend/"+projectName),
						strings.Join(packages, " "))); err != nil {
					helper.Log.Warn("npm install failed", "framework", framework, "error", err)
				}
			}
		}

		if framework == "React" {
			if err := helper.ScaffoldTemplate(ctx, containerID, "React", "/app/workspace/frontend/"+projectName); err != nil {
				helper.Log.Warn("React template scaffold failed", "error", err)
			}
			result.Pages = append(result.Pages, structs.Page{Route: "/", File: "src/App.jsx"})
		}
	}

	for _, framework := range backendServices {
		cfg, ok := configsMap[framework]
		if !ok {
			continue
		}

		if framework == "FastAPI" {
			if err := helper.ScaffoldTemplate(ctx, containerID, "FastAPI", "/app/workspace/backend"); err != nil {
				helper.Log.Warn("FastAPI template scaffold failed", "error", err)
			}

			if _, err := helper.ExecInContainer(ctx, containerID,
				"mkdir -p /app/workspace/backend/routers && touch /app/workspace/backend/routers/__init__.py"); err != nil {
				helper.Log.Warn("FastAPI routers directory failed", "error", err)
			}

			result.Endpoints = append(result.Endpoints, structs.Endpoint{
				Method: "GET",
				Path:   "/api/health",
				File:   "main.py",
			})
		}

		if cfg.ScaffoldCommand != nil {
			cmd := strings.ReplaceAll(*cfg.ScaffoldCommand, "{name}", projectName)
			helper.Log.Info("scaffolding backend", "framework", framework, "cmd", cmd)

			if _, err := helper.ExecInContainer(ctx, containerID, cmd); err != nil {
				helper.Log.Warn("backend scaffold failed", "framework", framework, "error", err)
			}
		}
	}

	for _, framework := range dbServices {
		cfg, ok := configsMap[framework]
		if !ok || cfg.ScaffoldCommand == nil {
			continue
		}

		cmd := strings.ReplaceAll(*cfg.ScaffoldCommand, "{name}", projectName)
		helper.Log.Info("scaffolding database", "framework", framework, "cmd", cmd)

		if _, err := helper.ExecInContainer(ctx, containerID,
			fmt.Sprintf("cd /app/workspace/database && %s", cmd)); err != nil {
			helper.Log.Warn("database scaffold failed", "framework", framework, "error", err)
		}
	}

	return result, nil
}

func CreateProjectContainer(
	ctx context.Context,
	db *sql.DB,
	builder *Builder,
	helper *ContainerHelper,
	projectID, projectName string,
	backendServices, frontendServices, dbServices []string,
	importURL string,
	env map[string]string,
) (*CreateProjectResult, error) {
	cfg, err := ProjectServicesConfig(ctx, db, builder, projectName, backendServices, frontendServices, dbServices)
	if err != nil {
		return nil, fmt.Errorf("resolving project config: %w", err)
	}

	containerResult, err := helper.CreateAndStartContainer(
		ctx, projectID, projectName, cfg.BaseTag, cfg.BaseType, cfg.CleanName,
		cfg.FrontendPort, backendServices, frontendServices, dbServices, env,
	)
	if err != nil {
		return nil, fmt.Errorf("creating container: %w", err)
	}

	var scan *structs.ScanResult
	var fresh *structs.ScaffoldResult

	if importURL != "" {
		scan, err = scaffoldImport(ctx, helper, containerResult.ContainerID, importURL)
		if err != nil {
			return nil, fmt.Errorf("scaffolding imported project: %w", err)
		}
	} else {
		fresh, err = scaffoldFresh(
			ctx, helper, containerResult.ContainerID, cfg.CleanName,
			frontendServices, backendServices, dbServices, cfg.ConfigsMap,
		)
		if err != nil {
			return nil, fmt.Errorf("scaffolding fresh project: %w", err)
		}
	}

	if err := helper.ScaffoldTemplate(ctx, containerResult.ContainerID, "LoggingService", "/"); err != nil {
		helper.Log.Warn("failed to scaffold LoggingService", "project_id", projectID, "error", err)
	}

	if err := helper.ExecDetached(ctx, containerResult.ContainerID, "logd > /var/log/logd.log 2>&1"); err != nil {
		helper.Log.Warn("failed to start logd", "project_id", projectID, "error", err)
	}

	result := &CreateProjectResult{
		ProjectID:   projectID,
		ContainerID: containerResult.ContainerID,
		Port:        containerResult.Port,
		URL:         containerResult.URL,
		BaseType:    cfg.BaseType,
		CleanName:   cfg.CleanName,
	}

	if scan != nil {
		repoName := strings.TrimSuffix(path.Base(strings.TrimSuffix(importURL, "/")), ".git")

		result.FrontendRoot = scan.FrontendRoot
		if result.FrontendRoot == "" {
			result.FrontendRoot = "/app/workspace/" + repoName
		}

		result.BackendRoot = scan.BackendRoot
		result.DBRoot = scan.DBRoot
		result.Pages = scan.Pages
		result.Endpoints = scan.Endpoints
		result.DetectedFrameworks = compactStrings(
			scan.FrontendFramework,
			scan.BackendFramework,
			scan.DBFramework,
		)
	} else {
		if len(frontendServices) > 0 {
			result.FrontendRoot = "/app/workspace/frontend/" + cfg.CleanName
		}
		if len(backendServices) > 0 {
			result.BackendRoot = "/app/workspace/backend"
		}
		if len(dbServices) > 0 {
			result.DBRoot = "/app/workspace/database"
		}

		result.Pages = fresh.Pages
		result.Endpoints = fresh.Endpoints
	}

	result.Groups = buildProjectGroups(
		ctx, helper, containerResult.ContainerID, cfg.CleanName,
		frontendServices, backendServices, scan,
	)

	if err := helper.StopContainer(ctx, containerResult.ContainerID); err != nil {
		helper.Log.Warn("failed to stop container after setup", "project_id", projectID, "error", err)
	}

	if _, err := db.ExecContext(ctx,
		`UPDATE projects SET container_id = $1 WHERE project_id = $2`,
		containerResult.ContainerID, projectID,
	); err != nil {
		return nil, fmt.Errorf("updating container_id: %w", err)
	}

	return result, nil
}
